### PR TITLE
 UCT/IB: fix uct_ep_flush with UCT_FLUSH_FLAG_CANCEL 

### PR DESCRIPTION
**Pull request by evgeny-leksikov**
## What
Invoke error handling flow for UCT_FLUSH_FLAG_CANCEL instead of just SW resources cleanup.

## Why ?
Fixes the bug which was rare reproducible on zcopy or RMA operation when user's buffer
is freed before operation completion.

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9478/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9508/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9516/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
@alinask pls take a look

File: src/uct/base/uct_iface.c [hunk 1]
```diff
@@ -363,11 +363,13 @@ ucs_status_t uct_set_ep_failed(ucs_class_t *cls, uct_ep_h tl_ep,
 
     if (iface->err_handler) {
         return iface->err_handler(iface->err_handler_arg, tl_ep, status);
+    } else if (status == UCS_ERR_CANCELED) {
+        status = UCS_OK;
```
Comment (alinask) [inline src/uct/base/uct_iface.c]: why isn't err handling invoked from here?
it should only be called from flush?

Comment (evgeny-leksikov) [reply src/uct/base/uct_iface.c]: If error handler is set, it will be invoked above, otherwise the error is not handled and this causes fatal error on transport level. Since the cancelation was initiated by user, we may assume that it's OK.

File: src/uct/ib/mlx5/ib_mlx5_log.c [hunk 2]
```diff
@@ -61,6 +61,10 @@ ucs_status_t uct_ib_mlx5_completion_with_err(uct_ib_iface_t *iface,
     uint32_t     qp_num;
     void         *wqe;
 
+    if (ecqe == NULL) {
+        return UCS_OK;
```
Comment (alinask) [inline src/uct/ib/mlx5/ib_mlx5_log.c]: can you explain this line please?

File: src/uct/base/uct_iface.c (hunk 1 above)
Comment (alinask) [reply src/uct/base/uct_iface.c]: so we dont care about any error flows that happened prior the user's cancellation?
why is error handling invoked then here https://github.com/openucx/ucx/pull/3408/files#diff-73e0f9c930e2212de847a8d24c0c0e05R53 ?

File: src/uct/ib/mlx5/ib_mlx5_log.c (hunk 2 above)
Comment (evgeny-leksikov) [reply src/uct/ib/mlx5/ib_mlx5_log.c]: Will comment the code, thanks

File: src/uct/base/uct_iface.c (hunk 1 above)
Comment (evgeny-leksikov) [reply src/uct/base/uct_iface.c]: > so we dont care about any error flows that happened prior the user's cancellation?

sorry, don't understand how it's related.. prior errors will be handled prior, right?

> why is error handling invoked then here 

to cleanup SW and HW resources.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9603/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
@yosefe pls take a look

File: src/uct/ib/dc/dc_mlx5_ep.c [hunk 3]
```diff
@@ -1224,3 +1222,64 @@ ucs_status_t uct_dc_mlx5_ep_check_fc(uct_dc_mlx5_iface_t *iface, uct_dc_mlx5_ep_
...
+                                                 status);
+        if (ep_status != UCS_OK) {
+            uct_ib_mlx5_completion_with_err(ib_iface, arg,
+                                            &iface->tx.dci_wqs[dci],
+                                            UCS_LOG_LEVEL_FATAL);
+            return;
+        }
+    }
+
+    uct_ib_mlx5_completion_with_err(ib_iface, arg, &iface->tx.dci_wqs[dci],
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5_ep.c]: are we going to print completion with err in case of cancel?

File: src/uct/ib/mlx5/ib_mlx5_log.c (hunk 2 above)
Comment (yosefe) [reply src/uct/ib/mlx5/ib_mlx5_log.c]: need to avoid calling log function from uct_dc_mlx5_ep_handle_failure
uct_rc_mlx5_iface_commom_clean, called from uct_dc_mlx5_iface_reset_dci, should remove all err completions during reset

File: src/uct/ib/rc/accel/rc_mlx5_iface.c [hunk 4]
```diff
@@ -170,20 +170,44 @@ static ucs_status_t uct_rc_mlx5_iface_arm_cq(uct_ib_iface_t *ib_iface,
 #endif
 }
 
+static ucs_status_t
+uct_rc_mlx5_ep_handle_failure(uct_rc_mlx5_ep_t *ep, ucs_status_t status)
```
Comment (yosefe) [inline src/uct/ib/rc/accel/rc_mlx5_iface.c]: shouldn't this be in rc_mlx5_ep.c?

File: src/uct/ib/rc/accel/rc_mlx5_iface.c [hunk 5]
```diff
@@ -170,20 +170,44 @@ static ucs_status_t uct_rc_mlx5_iface_arm_cq(uct_ib_iface_t *ib_iface,
...
+    uct_rc_iface_t     *iface;
+    uct_rc_mlx5_ep_t   *ep;
     uct_ib_mlx5_txwq_t txwq_copy;
-    size_t            txwq_size;
-    ucs_status_t      ep_status;
+    size_t             txwq_size;
+    unsigned           qp_num;
+    ucs_log_level_t    log_lvl;
 
+    if (status == UCS_ERR_CANCELED) {
```
Comment (yosefe) [inline src/uct/ib/rc/accel/rc_mlx5_iface.c]: need to put the code block which handles error reporting under if (status != CANCELED), so uct_rc_mlx5_ep_handle_failure() will be called only from one place in the code
- same comment for uct_rc_verbs_handle_failure

File: test/gtest/uct/test_flush.cc [hunk 6]
```diff
@@ -396,7 +393,7 @@ void uct_flush_test::test_flush_am_pending(flush_func_t flush, bool destroy_ep)
          } else {
             EXPECT_EQ(1, reqs.back().comp.count);
          }
-         reqs.pop_back();
+        reqs.pop_back();
```
Comment (yosefe) [inline test/gtest/uct/test_flush.cc]: why?

File: src/uct/ib/rc/accel/rc_mlx5_iface.c (hunk 4 above)
Comment (evgeny-leksikov) [reply src/uct/ib/rc/accel/rc_mlx5_iface.c]: then it must be non-static.. This is not a fast path, but I don't see the reason to increase @PLT when we can avoid this.

File: src/uct/ib/rc/accel/rc_mlx5_iface.c (hunk 5 above)
Comment (evgeny-leksikov) [reply src/uct/ib/rc/accel/rc_mlx5_iface.c]: this will add several branches in this function and make code hard readable.

File: test/gtest/uct/test_flush.cc (hunk 6 above)
Comment (evgeny-leksikov) [reply test/gtest/uct/test_flush.cc]: leftover..

File: src/uct/ib/dc/dc_mlx5_ep.c [hunk 7]
```diff
@@ -521,23 +521,21 @@ ucs_status_t uct_dc_mlx5_ep_get_zcopy(uct_ep_h tl_ep, const uct_iov_t *iov, size
...
     uct_dc_mlx5_ep_t    *ep    = ucs_derived_of(tl_ep, uct_dc_mlx5_ep_t);
     ucs_status_t        status;
     UCT_DC_MLX5_TXQP_DECL(txqp, txwq);
 
     if (ucs_unlikely(flags & UCT_FLUSH_FLAG_CANCEL)) {
-        if (ep->dci != UCT_DC_MLX5_EP_NO_DCI) {
-            uct_rc_txqp_purge_outstanding(&iface->tx.dcis[ep->dci].txqp,
-                                          UCS_ERR_CANCELED, 0);
-#if ENABLE_ASSERT
-            iface->tx.dcis[ep->dci].flags |= UCT_DC_DCI_FLAG_EP_CANCELED;
```
[the remark this answers is not in the document]
Comment (evgeny-leksikov) [reply src/uct/ib/dc/dc_mlx5_ep.c]: IIRC, it was needed to synchronize SW and HW counters, since we invoke error handling flow, it should work automatically by CQEs, will delete it.

[the remark this answers is not in the document]
Comment (yosefe) [reply src/uct/ib/dc/dc_mlx5_ep.c]: actually, it's used for assertion:
```
            /* The EP was destroyed after flush cancel */
            ucs_assert(ucs_test_all_flags(iface->tx.dcis[dci].flags,
                                          (UCT_DC_DCI_FLAG_EP_CANCELED |
                                           UCT_DC_DCI_FLAG_EP_DESTROYED)));
```

File: src/uct/ib/rc/accel/rc_mlx5_iface.c (hunk 4 above)
Comment (yosefe) [reply src/uct/ib/rc/accel/rc_mlx5_iface.c]: "don't see the reason to increase @plt when we can avoid this" -> might as well put all functions in one file..
if we decided to put all functions whose subject is "ep" to rc_mlx5_ep.c - should keep this order

File: src/uct/ib/dc/dc_mlx5_ep.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (evgeny-leksikov) [reply src/uct/ib/dc/dc_mlx5_ep.c]: Yes and this code was reachable because we were able to get CQEs after ep_flush(cancel) + ep_destroy

File: src/uct/ib/rc/accel/rc_mlx5_iface.c (hunk 4 above)
Comment (evgeny-leksikov) [reply src/uct/ib/rc/accel/rc_mlx5_iface.c]: ok, will fix

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9629/ for details (Mellanox internal link).

File: src/uct/ib/dc/dc_mlx5_ep.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (yosefe) [reply src/uct/ib/dc/dc_mlx5_ep.c]: 👍

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9636/ for details (Mellanox internal link).

File: src/uct/ib/rc/accel/rc_mlx5_iface.c (hunk 5 above)
Comment (yosefe) [reply src/uct/ib/rc/accel/rc_mlx5_iface.c]: @evgeny-leksikov on the other hand, after doing this, you will see the bug in your code (also in verbs flow)

Comment (evgeny-leksikov) [reply src/uct/ib/rc/accel/rc_mlx5_iface.c]: changed, but seems like i'm missing something, could you clarify pls?

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9678/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9684/ for details (Mellanox internal link).

File: src/uct/ib/rc/accel/rc_mlx5_iface.c (hunk 5 above)
Comment (yosefe) [reply src/uct/ib/rc/accel/rc_mlx5_iface.c]: the code is passing wc/cqe to a function which expects ep:
```
uct_rc_mlx5_ep_handle_failure(arg, status);
```
vs
```
if (uct_rc_mlx5_ep_handle_failure(ep, status) == UCS_OK) {
```
(AFAIR there was similar bug in UD transport)

File: src/uct/ib/dc/dc_mlx5_ep.c [hunk 8]
```diff
@@ -521,23 +521,22 @@ ucs_status_t uct_dc_mlx5_ep_get_zcopy(uct_ep_h tl_ep, const uct_iov_t *iov, size
...
-#endif
+        uct_ep_pending_purge(tl_ep, NULL, 0);
+        if (ep->dci == UCT_DC_MLX5_EP_NO_DCI) {
+            /* No dci -> no WQEs -> HW is clean, nothing to cancel */
+            return UCS_OK;
         }
 
-        uct_ep_pending_purge(tl_ep, NULL, 0);
+        uct_dc_mlx5_ep_handle_failure(ep, NULL, UCS_ERR_CANCELED);
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5_ep.c]: should we return NOT SUPPORTED if DCI policy is rand?

File: src/uct/ib/rc/accel/rc_mlx5_ep.c [hunk 9]
```diff
@@ -940,6 +946,29 @@ static UCS_CLASS_CLEANUP_FUNC(uct_rc_mlx5_ep_t)
...
+    uct_rc_txqp_purge_outstanding(&ep->super.txqp, status, 0);
+    /* poll_cqe for mlx5 returns NULL in case of failure and the cq_avaialble
+       is not updated for the error cqe and all outstanding wqes*/
+    rc_iface->tx.cq_available += ep->tx.wq.bb_max -
+                                 uct_rc_txqp_available(&ep->super.txqp);
+    return ib_iface->ops->set_ep_failed(ib_iface, &ep->super.super.super,
+                                        status);
+}
+
+ucs_status_t uct_rc_mlx5_ep_set_failed(uct_ib_iface_t *iface, uct_ep_h ep,
```
Comment (yosefe) [inline src/uct/ib/rc/accel/rc_mlx5_ep.c]: 👍 for moving it

File: src/uct/ib/rc/accel/rc_mlx5_iface.c [hunk 10]
```diff
@@ -174,16 +174,19 @@ static void
...
+    uct_rc_iface_t     *iface;
+    uct_rc_mlx5_ep_t   *ep;
     uct_ib_mlx5_txwq_t txwq_copy;
-    size_t            txwq_size;
-    ucs_status_t      ep_status;
+    size_t             txwq_size;
+    unsigned           qp_num;
+    ucs_log_level_t    log_lvl;
 
+    iface  = ucs_derived_of(ib_iface, uct_rc_iface_t);
```
Comment (yosefe) [inline src/uct/ib/rc/accel/rc_mlx5_iface.c]: why not do these initialization along with variable declaration, like before this PR?

File: src/uct/ib/rc/verbs/rc_verbs_iface.c [hunk 11]
```diff
@@ -44,24 +44,19 @@ static ucs_config_field_t uct_rc_verbs_iface_config_table[] = {
 static void uct_rc_verbs_handle_failure(uct_ib_iface_t *ib_iface, void *arg,
                                         ucs_status_t status)
 {
-    struct ibv_wc     *wc      = arg;
     uct_rc_iface_t    *iface   = ucs_derived_of(ib_iface, uct_rc_iface_t);
     ucs_log_level_t    log_lvl = UCS_LOG_LEVEL_FATAL;
+    struct ibv_wc     *wc;
     uct_rc_verbs_ep_t *ep;
 
+    wc = arg;
```
Comment (yosefe) [inline src/uct/ib/rc/verbs/rc_verbs_iface.c]: why this change?

File: test/gtest/uct/test_flush.cc [hunk 12]
```diff
@@ -525,13 +498,8 @@ UCS_TEST_P(uct_flush_test, am_flush_ep_nb) {
     m_flush_flags = UCT_FLUSH_FLAG_LOCAL;
     if (is_caps_supported(UCT_IFACE_FLAG_ERRHANDLE_PEER_FAILURE)) {
         test_flush_am_disconnect(&uct_flush_test::flush_ep_nb, false);
-
-        am_rx_count   = 0;
+        am_rx_count    = 0;
         m_flush_flags |= UCT_FLUSH_FLAG_CANCEL;
-        test_flush_am_disconnect(&uct_flush_test::flush_ep_nb, false);
```
Comment (yosefe) [inline test/gtest/uct/test_flush.cc]: maybe test_flush_am_pending() should assert that if flags have CANCEL, destroy_ep parameter must be true?

Comment (evgeny-leksikov) [reply test/gtest/uct/test_flush.cc]: good point, I think you meant `test_flush_am_disconnect`, we have several flush functions in the test

File: src/uct/ib/rc/verbs/rc_verbs_iface.c (hunk 11 above)
Comment (evgeny-leksikov) [reply src/uct/ib/rc/verbs/rc_verbs_iface.c]: left overs from previous implementation, will revert

File: test/gtest/uct/test_flush.cc [hunk 13]
```diff
@@ -401,6 +403,7 @@ void uct_flush_test::test_flush_am_pending(flush_func_t flush, bool destroy_ep)
      }
 
      if (destroy_ep) {
+        ASSERT_FALSE(is_flush_cancel());
```
Comment (yosefe) [inline test/gtest/uct/test_flush.cc]: should be other way around
if we flush with cancel - destroy_ep must be true
otherwise - destroy_ep could be either true or false

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9818/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9829/ for details (Mellanox internal link).

File: test/gtest/uct/test_flush.cc [hunk 14]
```diff
@@ -518,7 +524,7 @@ UCS_TEST_P(uct_flush_test, am_pending_flush_nb) {
         m_flush_flags |= UCT_FLUSH_FLAG_CANCEL;
     }
 
-    test_flush_am_pending(&uct_flush_test::flush_ep_nb, false);
+    test_flush_am_pending(&uct_flush_test::flush_ep_nb, true);
```
Comment (yosefe) [inline test/gtest/uct/test_flush.cc]: why are we always destroying ep now?
IMHO we should pass 'true' here only if we use CANCEL flag

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9855/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
@yosefe squash & merge?

**Comment by yosefe** [comment]
@evgeny-leksikov yes pls

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9868/ for details (Mellanox internal link).
